### DISCUSSION TITLE
How to decode ContractEmitted event data.

### DISCUSSION THREAD

**Discussion by blairmunroakusa** [polkascan/py-substrate-interface#354] [Q&A]
Hello! --discussion title speaks for itself.

I am struggling to understand how do decode the 'data' field for a ContractEmitted event. I have been running circles around scalecodec classes, substrate-interface ContractEvent, src, etc. to little avail.

Can somebody point me toward a resource/reference implementation/etc for decoding event metadata + values, one that is a little more explanatory than the scalecodec docs, substrate-interface docs, or the Substrate docs SCALE description?

Perhaps this question is better suited for the python scalecodec repo? I am at a loss.

Thank you for the consideration!

**Comment by arjanz** [comment]
The SCALE-bytes in the `data` field in a `ContractEmitted` is decoded according to the metadata of the contract. An example of such a metadata file (which is generated after [compiling the contract](https://use.ink/getting-started/building-your-contract)) is: https://github.com/polkascan/py-substrate-interface/blob/5f34d540982b2faf6f8f9c0fb41b027a2bba4f66/examples/assets/flipper.json#L49

To help decoding these ContractEvents, there are [Contract abstraction classes](https://polkascan.github.io/py-substrate-interface/examples/#create-and-call-ink-contract) available, for example the `ContractExecutionReceipt`: https://polkascan.github.io/py-substrate-interface/reference/contracts/#substrateinterface.contracts.ContractExecutionReceipt.__init__

This retrieves and decodes the extrinsic, events and ContractEvents. If you want to know more details how this data is decoded, have a look at this part of the code: 

https://github.com/polkascan/py-substrate-interface/blob/master/substrateinterface/contracts.py#L487 

and 

https://github.com/polkascan/py-substrate-interface/blob/master/substrateinterface/contracts.py#L412

Here the SCALE data is being decoded for the types defined in the contract metadata. I understand this is a tedious proces, there are no explicit delimiters or other context available in the data when decoding SCALE, which makes it efficient, but hard to debug or trace. 

The events of Substrate itself are basically decoded in the same way, this is decoded according to the `System.Events` SCALE-type defined in the metadata of the runtime.

**Reply by blairmunroakusa** [reply]
Thank you, will chew and report back.
